### ISSUE TITLE
Transform.scale animation bug - Icon won't re render on controller reset when value is zero

### ISSUE THREAD

**Issue by naamapps** [flutter/flutter#67259]
## Problem Description
When running a scale animation on an Icon widget, it works only once, but when reseting the controller to run the animation again it just wont re render the icon after the scale value has been zero.
If the end value is 0.01 it works as expected.

**Flutter Doctor:** 
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 1.22.0, on Mac OS X 10.15.6 19G2021, locale en-IL)
 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
[✓] Xcode - develop for iOS and macOS (Xcode 12.0.1)
[✓] Android Studio (version 4.0)
[✓] VS Code (version 1.49.3)
[✓] Connected device (1 available)

• No issues found!

## Sample code
```
import 'package:flutter/material.dart';
import 'package:flutter_sequence_animation/flutter_sequence_animation.dart';

class HeartOverlayAnimator extends StatefulWidget {
  final Stream<void> triggerAnimationStream;

  HeartOverlayAnimator({@required this.triggerAnimationStream});

  @override
  _HeartOverlayAnimatorState createState() => _HeartOverlayAnimatorState();
}

class _HeartOverlayAnimatorState extends State<HeartOverlayAnimator>
    with SingleTickerProviderStateMixin {
  AnimationController _controller;
  SequenceAnimation _sequenceAnimation;

  @override
  void initState() {
    super.initState();
    _controller = AnimationController(vsync: this);

    _sequenceAnimation = SequenceAnimationBuilder()
        .addAnimatable(
            animatable: Tween<double>(begin: 0.00, end: 80),
            from: const Duration(milliseconds: 0),
            to: const Duration(milliseconds: 800),
            tag: 'grow',
            curve: Curves.elasticOut)
        .addAnimatable(
            animatable: Tween<double>(begin: 80, end: 0.00),
            from: const Duration(milliseconds: 800),
            to: const Duration(milliseconds: 1200),
            tag: 'grow',
            curve: Curves.decelerate)
        .animate(_controller);

    widget.triggerAnimationStream.listen((_) {
      _controller
        ..reset()
        ..forward();
    });
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return AnimatedBuilder(
        animation: _controller,
        builder: (context, child) {
          return Transform.scale(
            scale: _sequenceAnimation['grow'].value,
            child: Icon(Icons.favorite, size: 1, color: Colors.white),
          );
        });
  }
}
```

**Comment by darshankawar** [comment]
@naamapps 
`flutter_sequence_animation` is a 3rd party plugin. Can you try by removing that plugin code and see if you still experience the issue ?
If so, please provide a code sample that doesn't have 3rd party code that shows the issue.
Thanks.

**Comment by naamapps** [comment]
Yeah no problem.
I tried without any third party plugin as well. still the same.

```
import 'package:flutter/material.dart';

class HeartOverlayAnimator extends StatefulWidget {
  final Stream<void> triggerAnimationStream;

  HeartOverlayAnimator({@required this.triggerAnimationStream});

  @override
  _HeartOverlayAnimatorState createState() => _HeartOverlayAnimatorState();
}

class _HeartOverlayAnimatorState extends State<HeartOverlayAnimator>
    with SingleTickerProviderStateMixin {
  AnimationController _heartController;
  Animation<double> _heartAnimation;

  @override
  void initState() {
    super.initState();
    final quick = const Duration(milliseconds: 500);
    final scaleTween = Tween(begin: 0.0, end: 1.0);
    _heartController = AnimationController(duration: quick, vsync: this);
    _heartAnimation = scaleTween.animate(
      CurvedAnimation(
        parent: _heartController,
        curve: Curves.elasticOut,
      ),
    );
    _heartController.addStatusListener((AnimationStatus status) {
      if (status == AnimationStatus.completed) {
        _heartController.animateTo(
          0.0,
          duration: const Duration(milliseconds: 300),
          curve: Curves.decelerate,
        );
      }
    });

    widget.triggerAnimationStream.listen((_) {
      _heartController
        ..reset()
        ..forward();
    });
  }

  @override
  void dispose() {
    _heartController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return ScaleTransition(
      scale: _heartAnimation,
      child: Icon(Icons.favorite, size: 80.0, color: Colors.white),
    );
  }
}
```

**Comment by darshankawar** [comment]
@naamapps 
Not sure what parameter value I need to pass here:

`HeartOverlayAnimator(triggerAnimationStream: ),`

**Comment by naamapps** [comment]
@darshankawar 

The way I have set it up is that I have a stream controller:
`final StreamController<void> streamController = StreamController.broadcast(); `

I need it because there are two widgets in my widget tree who need this stream events.

On double tap on an image I call:
`streamController.sink.add(null);`

Two of the widgets that need the stream have a listener inside of them to play an animation when the event fires.
The first one is working fine (the animation doesn't go to scale zero, so it's fine - But if the end value is zero, it has the same bug).
The other one (`HeartOverlayAnimator`) does get the events as expected but have an issue with the animation when the end value is zero.

I hope the explanation is clear. 
Thanks

**Comment by darshankawar** [comment]
@naamapps 
So the issue is, the  `Icon(Icons.favorite, size: 80.0, color: Colors.white),` isn't rendering with above code, correct ?
I see a black screen when I run.
Thanks.

**Comment by naamapps** [comment]
@darshankawar

Not exactly.
As I said, I have the stream to get an event when the user double tap an image, and when I get the event, the heart overlay should appear with an animation on the image (like instagram).

The initial view **should** be invisible (like the view you have).
Only when the user double tap, it would appear for the first time.
But if the user double tap again - it won't appear again.

**That is the bug - the heart won't re render after the first animation if the animation's end value is zero.**

**Comment by darshankawar** [comment]
If I double tap on invisible view, I get nothing still.

**Comment by naamapps** [comment]
@darshankawar
Maybe try on an image like so:

```
import 'dart:async';
import 'package:flutter/material.dart';
import 'package:testApp/widgets/heart_overlay_animator.dart';

class Test extends StatelessWidget {
  const Test({Key key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    StreamController<void> streamController = StreamController.broadcast();
    return GestureDetector(
      onDoubleTap: () => streamController.sink.add(null),
      child: Stack(
        alignment: Alignment.center,
        children: [
          Positioned.fill(
            child: ClipRRect(
              borderRadius: BorderRadius.circular(8.0),
              child: Image.network(
                'https://cdn.pixabay.com/photo/2020/10/01/09/35/bee-5618012_1280.jpg',
                fit: BoxFit.cover,
              ),
            ),
          ),
          HeartOverlayAnimator(
            triggerAnimationStream: streamController.stream,
          ),
        ],
      ),
    );
  }
}
```

**Comment by darshankawar** [comment]
@naamapps 
Tried the updated code sample on latest stable and on Android 8.1.0 device. Double tapping on screen shows the heart animation everytime. (ignore the red screen, since I didn't wrap in Material widget).

[67259.mp4.zip](https://github.com/flutter/flutter/files/5338522/67259.mp4.zip)

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel stable, 1.22.0, on Mac OS X 10.15.4 19E2269, locale en-IN)
    • Flutter version 1.22.0 at /Users/dhs/documents/Fluttersdk/flutter
    • Framework revision d408d302e2 (7 days ago), 2020-09-29 11:49:17 -0700
    • Engine revision 5babba6c4d
    • Dart version 2.10.0

 
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.0)
    • Android SDK at /Users/dhs/Library/Android/sdk
    • Platform android-30, build-tools 30.0.0
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[!] Xcode - develop for iOS and macOS (Xcode 12.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.0.1, Build version 12A7300
    ✗ Xcode requires additional components to be installed in order to run.
      Launch Xcode and install additional required components when prompted or
      run:
        sudo xcodebuild -runFirstLaunch
    • CocoaPods version 1.9.3

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 46.0.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.49.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.11.0

[!] Connected device
    ! No devices available

! Doctor found issues in 2 categories.
```

</details>

**Comment by naamapps** [comment]
@darshankawar 
Please try it with an actual icon, because in other widgets (such as images) the animation does re render.
The red screen is not an icon, so it renders every time.

I did a little testing and found out the text and icons won't re render.
Maybe issue #29440 came back?

**Comment by darshankawar** [comment]
Tried 29440 and it seems to be working properly by rendering text and icon together.

[67259.mp4.zip](https://github.com/flutter/flutter/files/5345587/67259.mp4.zip)

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel beta, 1.22.0, on Mac OS X 10.15.4 19E2269, locale en-IN)
    • Flutter version 1.22.0 at /Users/dhs/documents/Fluttersdk/flutter
    • Framework revision d408d302e2 (8 days ago), 2020-09-29 11:49:17 -0700
    • Engine revision 5babba6c4d
    • Dart version 2.10.0

 
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.0)
    • Android SDK at /Users/dhs/Library/Android/sdk
    • Platform android-30, build-tools 30.0.0
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.0.1, Build version 12A7300
    • CocoaPods version 1.9.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 46.0.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.49.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.11.0

[✓] Connected device (3 available)
    • SM A260G (mobile) • 5200763ebcfa861f • android-arm    • Android 8.1.0 (API
      27)
    • Web Server (web)  • web-server       • web-javascript • Flutter Tools
    • Chrome (web)      • chrome           • web-javascript • Google Chrome
      86.0.4240.75

• No issues found!
```

</details>

**Comment by naamapps** [comment]
@darshankawar 
So I tested a bit more and found out that repeating the animation infinitely by calling `_heartController.repeat();` is working fine and re renders everything, as you can see here:
[Repeating animation - rerenders.mp4.zip](https://github.com/flutter/flutter/files/5345864/Repeating.animation.-.rerenders.mp4.zip)

However if you double tap and run this
```
 _heartController
        ..reset()
        ..forward();
```
The first time it's working - all other times the text and icon won't re render as you can see here:
[Double tap animation - no rerender.mp4.zip](https://github.com/flutter/flutter/files/5345866/Double.tap.animation.-.no.rerender.mp4.zip)

Another finding is that if you double tap fast (before the animation ends) - it does re render.

I hope you can reproduce the bug this time.
Thanks

**Comment by erez213** [comment]
I have the same issue.
I use ScaleTransition, and on the first time the text is visible, but on the second time I can't see it.

**Comment by darshankawar** [comment]
@naamapps 
Saw something very [related](https://github.com/flutter/flutter/issues/67529) issue today that I was able to replicate on latest beta, but not on Master. Does it sound similar to yours?
Thanks.

**Comment by naamapps** [comment]
@darshankawar 
Yes it's exactly the same I think.
I ran it on my phone, and in this example it sometimes won't render even on the first time the animation runs.
The weird part here is that i'm on the latest stable (not beta), so I don't really know what to say about this.

Did you see the videos I attached in my last message?
I clearly shows that there is a bug.

I'm on Flutter 1.22.0 stable, running on Android 10 (Samsung galaxy note 20 ultra, if that matters).
Thanks.

**Comment by darshankawar** [comment]
Issue replicable on latest master. Tried on Android 8.1.0 device. Double tapping on the animation shows the text and icon for first time, but subsequent double taps don't render them.

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel master, 1.23.0-8.0.pre.221, on Mac OS X 10.15.4 19E2269
    x86_64, locale en-IN)
    • Flutter version 1.23.0-8.0.pre.221 at
      /Users/dhs/documents/Fluttersdk/flutter
    • Framework revision 0d8d4f77aa (23 hours ago), 2020-10-08 09:26:30 +0200
    • Engine revision 2eac514f26
    • Dart version 2.11.0 (build 2.11.0-186.0.dev)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.0)
    • Android SDK at /Users/dhs/Library/Android/sdk
    • Platform android-30, build-tools 30.0.0
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.0.1, Build version 12A7300
    • CocoaPods version 1.9.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 46.0.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.49.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.11.0

[✓] Connected device (4 available)
    • SM A260G (mobile) • 5200763ebcfa861f • android-arm    • Android 8.1.0 (API
      27)
    • macOS (desktop)   • macos            • darwin-x64     • Mac OS X 10.15.4
      19E2269 x86_64
    • Web Server (web)  • web-server       • web-javascript • Flutter Tools
    • Chrome (web)      • chrome           • web-javascript • Google Chrome
      86.0.4240.75

• No issues found!

```

</details>

**Comment by darshankawar** [comment]
May be related to [30522](https://github.com/flutter/flutter/issues/30522)

**Comment by niktses** [comment]
I am experiencing the same issue since updating to version 1.22. It was working fine before. I am always on stable.

In my case i use:

```
AnimatedBuilder(
              animation: animation,
              builder: (context, child) {
                return Transform(
                  transform: Matrix4.identity()
                    ..setEntry(2, 2, 0.005)
                    ..rotateY(animation.value),
                  origin: Offset(25, 0),
                  child: child,
                );
              },
              child: Icon(icon, size: 50, color: CustomColors.MainBlue),
            ),
```

And it dissapears when the icon is rotated to pi / 2.

It feels like when then there is nothing to render it optimizes for performance and drops the icon altogether but then it is never returned again.

**Comment by Ocelyn** [comment]
Same issue on 1.22.2 stable channel

**Comment by jason-simmons** [comment]
I tried reproducing this on the current master but was unable to.

This is the code that I used (based on the examples above):

<details>
  <summary>Code</summary>

```
import 'dart:async';
import 'package:flutter/material.dart';

void main() {
  runApp(Directionality(
    textDirection: TextDirection.ltr,
    child: Test()
  ));
}

class Test extends StatelessWidget {
  const Test({Key key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    StreamController<void> streamController = StreamController.broadcast();
    return GestureDetector(
      onDoubleTap: () => streamController.sink.add(null),
      child: Stack(
        alignment: Alignment.center,
        children: [
          Positioned.fill(
            child: ClipRRect(
              borderRadius: BorderRadius.circular(8.0),
              child: Image.network(
                'https://cdn.pixabay.com/photo/2020/10/01/09/35/bee-5618012_1280.jpg',
                fit: BoxFit.cover,
              ),
            ),
          ),
          HeartOverlayAnimator(
            triggerAnimationStream: streamController.stream,
          ),
        ],
      ),
    );
  }
}

class HeartOverlayAnimator extends StatefulWidget {
  final Stream<void> triggerAnimationStream;

  HeartOverlayAnimator({@required this.triggerAnimationStream});

  @override
  _HeartOverlayAnimatorState createState() => _HeartOverlayAnimatorState();
}

class _HeartOverlayAnimatorState extends State<HeartOverlayAnimator>
    with SingleTickerProviderStateMixin {
  AnimationController _heartController;
  Animation<double> _heartAnimation;

  @override
  void initState() {
    super.initState();
    final quick = const Duration(milliseconds: 500);
    final scaleTween = Tween(begin: 0.0, end: 1.0);
    _heartController = AnimationController(duration: quick, vsync: this);
    _heartAnimation = scaleTween.animate(
      CurvedAnimation(
        parent: _heartController,
        curve: Curves.elasticOut,
      ),
    );
    _heartController.addStatusListener((AnimationStatus status) {
      if (status == AnimationStatus.completed) {
        _heartController.animateTo(
          0.0,
          duration: const Duration(milliseconds: 300),
          curve: Curves.decelerate,
        );
      }
    });

    widget.triggerAnimationStream.listen((_) {
      _heartController
        ..reset()
        ..forward();
    });
  }

  @override
  void dispose() {
    _heartController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return ScaleTransition(
      scale: _heartAnimation,
      child: Icon(Icons.favorite, size: 80.0, color: Colors.white),
    );
  }
}
```
</details>

The animation rendered correctly each time I double tapped.  I tried this on Moto G4 and Nexus 5X devices.

@darshankawar If this is still happening on master, then can you post the app that you're using to reproduce it?

**Comment by darshankawar** [comment]
@jason-simmons Tried on latest master and not seeing the issue anymore.

@naamapps  can you check from your end if you are still seeing the issue ?

This is what I see now, which renders the icon properly everytime I double tap.

[67259.mp4.zip](https://github.com/flutter/flutter/files/5456528/67259.mp4.zip)

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel master, 1.24.0-4.0.pre.82, on Mac OS X 10.15.4 19E2269
    darwin-x64, locale en-IN)
    • Flutter version 1.24.0-4.0.pre.82 at
      /Users/dhs/documents/Fluttersdk/flutter
    • Framework revision e5814756a2 (2 days ago), 2020-10-27 01:47:03 -0700
    • Engine revision 1857470267
    • Dart version 2.11.0 (build 2.11.0-260.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.0)
    • Android SDK at /Users/dhs/Library/Android/sdk
    • Platform android-30, build-tools 30.0.0
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.0.1, Build version 12A7300
    • CocoaPods version 1.9.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 46.0.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.50.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.15.1

[✓] Connected device (4 available)
    • SM A260G (mobile) • 5200763ebcfa861f • android-arm    • Android 8.1.0 (API
      27)
    • macOS (desktop)   • macos            • darwin-x64     • Mac OS X 10.15.4
      19E2269 darwin-x64
    • Web Server (web)  • web-server       • web-javascript • Flutter Tools
    • Chrome (web)      • chrome           • web-javascript • Google Chrome
      86.0.4240.111

• No issues found!


```

</details>

**Comment by naamapps** [comment]
@darshankawar 
Can confirm the issue is resolved in latest master.
Thanks guys!

**Comment by darshankawar** [comment]
Closing this as resolved based on @naamapps's last message above.

Could everyone who still has this problem please file a new issue with the exact description of what happens, logs, and the output of `flutter doctor -v`.
All system setups can be slightly different, so it's always better to open new issues and reference related issues.
